### ISSUE TITLE
Programmer Editing Access New Idea (Priority Functionality)

### ISSUE THREAD

**Issue by galenjosephhunter** [wavefarm/site#165]
Rather than giving programmers direct admin access, can we create a login on the public site so that when programmers would go to a specific broadcast via https://wavefarm.org/wgxc/schedule they would click login to edit, and then have a very user friendly window allowing them to edit the broadcast name, description, and add image, feature, audio.

**Comment by gsf** [comment]
Should follow broadcast automation. Also move this to the admin repo.

**Comment by gsf** [comment]
Should be straightforward to add a form page within the schedule, triggered from individual broadcasts. A bigger issue is the requirement for authentication and some kind of login on the public site for these programmers.

**Comment by ptberry** [comment]
Brief Spec:

Implementation:

- Add "Staff" and "Programmer" roles to the existing "role" field on users, which will remain a simple select.
- Add a "users" relation to show to capture show level permissions; ideally this relation will only show up on user side when role=Programmer.
- Each show will get a new admin user with roll="Programmer"
- Because there will be a new, limited permission role we need to restrict most admin functionality to role="Staff" (see question below)
- We will keep the existing logon for now for all users, with possible post-logon redirect based on role. A possible future enhancement would be adding a new logon to the site itself so programmers wouldn't need to visit /admin at all.
- The programmer editing interface will actually live in the front end. Once logged on programmers will be able to access the interface via the public broadcast pages.

Questions:

1) Is it sufficient to lock down the relevant API routes (item-delete, item-set) to enforce roles or so we also need something in /admin ?
2) Is it OK for programmers to only have access to published broadcasts? Otherwise they will need an alternate way to reach unpublished broadcast pages.
3) Do we need to mark edited broadcasts somehow for Staff review ? Should edits be immediately published?
4) What exactly should the interface look like? A popup form? An inline form? Perhaps even in-line field editing?

**Comment by galenjosephhunter** [comment]
Not sure if I should be answering this or if it is geared toward Gabe, but my comments are:

Implementation sounds good. Responses to your Questions:

1) Note sure what that means in practical terms.
2) Yes, programmers should only have access to published broadcasts.
3) Yes. Lynn says that our hope is that programmer edited broadcasts will be cued for staff review before being published.
4) Ideally in-line field editing, though with number 3’s answer in the mix, I wonder about in-line because it might be confusing as to why programmers don’t see the changes go live immediately. Your thoughts?
> 1) Is it sufficient to lock down the relevant API routes (item-delete, item-set) to enforce roles or so we also need something in /admin ?
> 2) Is it OK for programmers to only have access to published broadcasts? Otherwise they will need an alternate way to reach unpublished broadcast pages.
> 3) Do we need to mark edited broadcasts somehow for Staff review ? Should edits be immediately published?
> 4) What exactly should the interface look like? A popup form? An inline form? Perhaps even in-line field editing?
>

**Comment by ptberry** [comment]
Yes let's talk it though here for now. 

1) For example, should programmers be able to search and view all objects in admin? Meaning perhaps they can view things besides broadcasts even though they can't edit them. The other extreme would be that that cannot see anything whatsoever in /admin and only have access to this new interface. The latter would require at least some work in /admin as well as the API.
2,3,4) I suggest we table any kind of approval type workflow for now. Perhaps we can get away with the changes going live immediately but flagging the broadcast for later review by staff? That would solve a few problems for now and keep this relatively simple.

**Comment by galenjosephhunter** [comment]
1) I’d rather programmers only have access to the new interface.
2,3,4) Sounds good to me.

**Comment by ptberry** [comment]
@gsf : "Spec" so far is above.

**Comment by gsf** [comment]
I agree with the decision to make changes live but keep a record for review. See my rather stilted thoughts on an across-the-board logging mechanism at https://github.com/wavefarm/api/issues/59.

I also agree that the programmers should just edit via the public interface. Building up a new interface for them would likely take around the same amount of time (or less) as modifying the admin to limit what they can see there.
